File: frontend/src/services/points.ts
import { apiClient } from "./api";
import type { AxiosError } from "axios";
import type { PointCreate, PointFinish, PointUpdate, PointWithPlayers } from "../types";

// Start a new point (creates an active point)
export const startPoint = async (data: PointCreate): Promise<PointWithPlayers> => {
  const response = await apiClient.post<PointWithPlayers>("/points", data);
  return response.data;
};

// Finish an active point
export const finishPoint = async (
  pointId: number,
  data: PointFinish
): Promise<PointWithPlayers> => {
  const response = await apiClient.post<PointWithPlayers>(`/points/${pointId}/finish`, data);
  return response.data;
};

// Cancel (delete) an active point
export const cancelPoint = async (pointId: number): Promise<void> => {
  await apiClient.delete(`/points/${pointId}/cancel`);
};

// Get active point for a game (ready or running) (returns null if 404)
export const getActivePoint = async (gameId: number): Promise<PointWithPlayers | null> => {
  try {
    const response = await apiClient.get<PointWithPlayers>(`/points/games/${gameId}/active`);
    return response.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    if (axiosError.response?.status === 404) {
      return null;
    }
    throw error;
  }
};

// Deprecated: Use getActivePoint instead
// Get running point for a game (returns null if 404)
export const getRunningPoint = async (gameId: number): Promise<PointWithPlayers | null> => {
  try {
    const response = await apiClient.get<PointWithPlayers>(`/points/games/${gameId}/running`);
    return response.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    if (axiosError.response?.status === 404) {
      return null;
    }
    throw error;
  }
};

// Get point by ID with player details
export const getPoint = async (pointId: number): Promise<PointWithPlayers> => {
  const response = await apiClient.get<PointWithPlayers>(`/points/${pointId}`);
  return response.data;
};

// Update point
export const updatePoint = async (
  pointId: number,
  data: PointUpdate
): Promise<PointWithPlayers> => {
  const response = await apiClient.put<PointWithPlayers>(`/points/${pointId}`, data);
  return response.data;
};

// Delete point
export const deletePoint = async (pointId: number): Promise<void> => {
  await apiClient.delete(`/points/${pointId}`);
};

File: backend/app/routers/points.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from app import schemas, crud
from app.auth.context import AccessContext
from app.auth.dependencies import get_request_access_context
from app.auth.redaction import serialize_point
from app.database import get_db
from app.logging_config import get_logger

logger = get_logger("routers.points")

router = APIRouter(
    prefix="/points",
    tags=["points"]
)


@router.post("", response_model=schemas.PointWithPlayers, status_code=201)
def create_point(point: schemas.PointCreate, db: Session = Depends(get_db)):
    try:
        created_point = crud.create_point(db, point)
        player_count = len(point.player_ids) if point.player_ids else 0
        logger.info(
            f"Point created: id={created_point.id}, game={point.game_id}, "
            f"offense={point.starting_on_offense}, players={player_count}, status=ready"
        )
        return created_point
    except ValueError as e:
        logger.warning(
            f"Failed to create point for game {point.game_id}: {str(e)}"
        )
        # Map known not-found errors to 404
        error_text = str(e).lower()
        status_code = 404 if error_text in ["game not found", "strategy not found"] else 400
        raise HTTPException(status_code=status_code, detail=str(e))


@router.get("/{point_id}", response_model=schemas.PointWithPlayers)
def get_point(
    point_id: int,
    db: Session = Depends(get_db),
    access_context: AccessContext = Depends(get_request_access_context),
):
    point = crud.get_point(db, point_id)
    if not point:
        raise HTTPException(status_code=404, detail="Point not found")
    return serialize_point(point, access_context)


@router.put("/{point_id}", response_model=schemas.PointWithPlayers)
def update_point(point_id: int, point_update: schemas.PointUpdate, db: Session = Depends(get_db)):
    try:
        point = crud.update_point(db, point_id, point_update)
        if not point:
            logger.warning(f"Failed to update point: point {point_id} not found")
            raise HTTPException(status_code=404, detail="Point not found")

        # Log status changes (important for tracking point lifecycle)
        if point_update.status:
            logger.info(f"Point {point_id} status changed to {point_update.status.value}")

        return point
    except ValueError as e:
        logger.warning(f"Failed to update point {point_id}: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{point_id}", status_code=204)
def delete_point(point_id: int, db: Session = Depends(get_db)):
    success = crud.delete_point(db, point_id)
    if not success:
        raise HTTPException(status_code=404, detail="Point not found")


@router.post("/{point_id}/finish", response_model=schemas.PointWithPlayers)
def finish_point(point_id: int, finish_data: schemas.PointFinish, db: Session = Depends(get_db)):
    try:
        point = crud.finish_point(db, point_id, finish_data)
        if not point:
            logger.warning(f"Failed to finish point: point {point_id} not found")
            raise HTTPException(status_code=404, detail="Point not found")

        logger.info(f"Point {point_id} finished: won={finish_data.won}")
        return point
    except ValueError as e:
        logger.warning(f"Failed to finish point {point_id}: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{point_id}/cancel", status_code=204)
def cancel_point(point_id: int, db: Session = Depends(get_db)):
    try:
        success = crud.cancel_point(db, point_id)
        if not success:
            logger.warning(f"Failed to cancel point: point {point_id} not found")
            raise HTTPException(status_code=404, detail="Point not found")

        logger.info(f"Point {point_id} cancelled")
    except ValueError as e:
        logger.warning(f"Failed to cancel point {point_id}: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/games/{game_id}/running", response_model=schemas.PointWithPlayers)
def get_running_point(
    game_id: int,
    db: Session = Depends(get_db),
    access_context: AccessContext = Depends(get_request_access_context),
):
    """Get the running point for a game"""
    # Verify game exists
    game = crud.get_game(db, game_id)
    if not game:
        raise HTTPException(status_code=404, detail="Game not found")

    point = crud.get_running_point_for_game(db, game_id)
    if not point:
        raise HTTPException(status_code=404, detail="No running point found for this game")
    return serialize_point(point, access_context)


@router.get("/games/{game_id}/active", response_model=schemas.PointWithPlayers)
def get_active_point(
    game_id: int,
    db: Session = Depends(get_db),
    access_context: AccessContext = Depends(get_request_access_context),
):
    """Get the active (ready or running) point for a game"""
    # Verify game exists
    game = crud.get_game(db, game_id)
    if not game:
        raise HTTPException(status_code=404, detail="Game not found")

    point = crud.get_active_point_for_game(db, game_id)
    if not point:
        raise HTTPException(status_code=404, detail="No active point found for this game")
    return serialize_point(point, access_context)
